crates/core_simd/examples/nbody.rs

```diff
@@ -0,0 +1,190 @@
+            -2.96589568540237556e-05 * DAYS_PER_YEAR,
+            0.,
+        ]),
+        mass: 4.36624404335156298e-05 * SOLAR_MASS,
+    },
+    // neptune:
+    Body {
+        x: f64x4::from_array([
+            1.53796971148509165e+01,
+            -2.59193146099879641e+01,
+            1.79258772950371181e-01,
+            0.,
+        ]),
+        v: f64x4::from_array([
+            2.68067772490389322e-03 * DAYS_PER_YEAR,
+            1.62824170038242295e-03 * DAYS_PER_YEAR,
+            -9.51592254519715870e-05 * DAYS_PER_YEAR,
+            0.,
+        ]),
+        mass: 5.15138902046611451e-05 * SOLAR_MASS,
+    },
+];
+
+pub fn offset_momentum(bodies: &mut [Body; N_BODIES]) {
+    let (sun, rest) = bodies.split_at_mut(1);
+    let sun = &mut sun[0];
+    for body in rest {
+        let m_ratio = body.mass / SOLAR_MASS;
+        sun.v -= body.v * m_ratio;
+    }
+}
+
+pub fn energy(bodies: &[Body; N_BODIES]) -> f64 {
+    let mut e = 0.;
+    for i in 0..N_BODIES {
+        let bi = &bodies[i];
+        e += bi.mass * (bi.v * bi.v).horizontal_sum() * 0.5;
+        for bj in bodies.iter().take(N_BODIES).skip(i + 1) {
+            let dx = bi.x - bj.x;
+            e -= bi.mass * bj.mass / (dx * dx).horizontal_sum().sqrt()
+        }
+    }
+    e
+}
+
+pub fn advance(bodies: &mut [Body; N_BODIES], dt: f64) {
+    const N: usize = N_BODIES * (N_BODIES - 1) / 2;
+
+    // compute distance between bodies:
+    let mut r = [f64x4::splat(0.); N];
+    {
+        let mut i = 0;
+        for j in 0..N_BODIES {
+            for k in j + 1..N_BODIES {
+                r[i] = bodies[j].x - bodies[k].x;
+                i += 1;
+            }
+        }
+    }
+
+    let mut mag = [0.0; N];
+    let mut i = 0;
+    while i < N {
+        let d2s = f64x2::from_array([
+            (r[i] * r[i]).horizontal_sum(),
+            (r[i + 1] * r[i + 1]).horizontal_sum(),
+        ]);
+        let dmags = f64x2::splat(dt) / (d2s * d2s.sqrt());
+        // dmags.write_to_slice_unaligned(&mut mag[i..]);
+        mag[i] = dmags[0];
+        mag[i + 1] = dmags[1];
+        i += 2;
+    }
+
+    i = 0;
+    for j in 0..N_BODIES {
+        for k in j + 1..N_BODIES {
+            let f = r[i] * mag[i];
+            bodies[j].v -= f * bodies[k].mass;
+            bodies[k].v += f * bodies[j].mass;
+            i += 1
+        }
+    }
+    for body in bodies {
+        body.x += dt * body.v
+    }
+}
+
+// #[inline]
+// pub unsafe fn write_to_slice_unaligned(slice: &mut SimdF64::<LANES>) {
+//     let target_ptr = slice.get_unchecked_mut(0) as *mut f64x2;
+//     *(target_ptr as *mut f64x2) = SimdF64;
+// }
+
+pub fn run(n: usize) -> (f64, f64) {
+    let mut bodies = BODIES;
+    offset_momentum(&mut bodies);
+    let energy_before = energy(&bodies);
+    for _ in 0..n {
+        advance(&mut bodies, 0.01);
+    }
+    let energy_after = energy(&bodies);
+
+    (energy_before, energy_after)
+}
+
+fn approx_eq_f32(a: f32, b: f32) -> bool {
```

Comment:
Note that when i used this as an example on zulip... this particular threshold is just "the smallest value where the intel rsqrt intrinsic seemed to stop complaining", so maybe we could change the threshold, but since this is just an example i think it's probably fine to pick a threshold out of thin air.

Reply:
I demoted the answers to f32 to make it work, so I don't mind fudging it again for the purposes of demonstration.